docs/docs/optimizing-performance.md

```diff
@@ -0,0 +1,228 @@
+
+```js
+class CounterButton extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {count: 1};
+  }
+
+  render() {
+    <button
+      color={this.props.color}
+      onClick={() => this.setState(state => ({count: state.count + 1}))}>
+      Count: {this.state.count}
+    </button>
+  }
+}
+```
+
+Most of the time, you can use `React.PureComponent` instead of writing your own `shouldComponentUpdate`. It only does a shallow comparison, so you can't use it when the rendering function uses deep properties of the props or state. For example, if your render function refers to `this.props.foo.bar` or `this.state.qux.quux`, you should not use `React.PureComponent`.
+
+If you create functions or child components in your `render` function, `PureComponent` will never short-circuit rendering, since those functions or child components will always be different. You can often avoid this by creating these once during component creation rather than on every render.
+
+## shouldComponentUpdate In Action
+
+Here's a subtree of components. For each one, `SCU` indicates what `shouldComponentUpdate` returned, and `vDOMEq` indicates whether the rendered DOM elements were equivalent. Finally, the circle's color indicates whether the component had to be reconciled or not.
+
+<figure><img src="/react/img/docs/should-component-update.png" /></figure>
+
+Since `shouldComponentUpdate` returned `false` for the subtree rooted at C2, React did not attempt to render C2, and thus didn't even have to invoke `shouldComponentUpdate` on C4 and C5.
+
+For C1 and C3, `shouldComponentUpdate` returned `true`, so React had to go down to the leaves and check them. For C6 `shouldComponentUpdate` returned `true`, and since the rendered DOM elements weren't equivalent React had to update the DOM.
+
+The last interesting case is C8. React had to render this component, but since the DOM elements it returned were equal to the previously rendered ones, it didn't have to update the DOM.
+
+Note that React only had to do DOM mutations for C6, which was inevitable. For C8, it bailed out by comparing the rendered DOM elements, and for C2's subtree and C7, it didn't even have to compare the elements as we bailed out on `shouldComponentUpdate`, and `render` was not called.
+
+## Tricky Examples
+
+Let's say that you have a component that just renders the string passed in `props.value`. We could implement this with `React.PureComponent`:
+
+```javascript
+class StringComponent extends React.PureComponent {
+  render() {
+    return <div>{this.props.value}</div>;
+  }
+}
+```
+
+If your components use more complex data structures, this might not be so simple. For example, let's say your component renders a comma-separated list of words. This code does *not* work correctly:
```

Comment:
OK - i added a second component. I also moved the diagram up, so that this section and the section that goes into detail about not mutating things are next to each other, because they are kind of starting to merge in terms of content.